### DISCUSSION TITLE
Help with dynamic routing

### DISCUSSION THREAD

**Discussion by Paultje52** [vercel/next.js#26336] [Help] [answered]
Hello. I want to use dynamic routing in my project. I have a class that extends `react.Component` and I export it using next/router by using `withRouter`. The filename is `[...page].js`.z
```js
class Page extends Component {
  constructor(props) {
    super(props);
    console.log(props.router.query.page);
  }
  // ....
}
export default withRouter(Page);
```

The problem is that when I reload the page, I get `undefined` in the console. To get the right page variable in my console, I have to change something in the constructor so the component gets "reloaded".

When reading the [`Caveats`](https://nextjs.org/docs/routing/dynamic-routes#caveats) section in the docs, I read this.
```
Pages that are statically optimized by Automatic Static Optimization will be hydrated without their route parameters provided, i.e query will be an empty object ({}).

After hydration, Next.js will trigger an update to your application to provide the route parameters in the query object.
```

But how do I get that update? I'm using next.js 11 and I'm in dev-mode (`npm run dev`).

**Comment by adebiyial** [comment, marked as answer]
Hey. Read this https://www.smashingmagazine.com/2021/06/client-side-routing-next-js/

**Reply by Paultje52** [reply]
Hello. Thank you for your reply. Your link helped me better understand dynamic routing, but I still have the same issue with that `props.router.query.page` is undefined.

**Reply by adebiyial** [reply]
You're welcome.

If you're using `router` [in a dynamic route], you need to be aware that it'd be undefined for a brief moment. Did you do a check to see if `query` exists on `router`?

**Reply by Paultje52** [reply]
`props.router.query` is an empty object. Do I need to wait and check if the object isn't empty anymore?

**Reply by adebiyial** [reply]
Yes, you have to.

**Reply by Paultje52** [reply]
That doesn't work either. I used `setTimeout` in the constructor. One timeout of 1s and one timeout of 10s, but both times the object is empty.
```js
    setTimeout(() => {
      console.log(2, props.router.query);
    }, 1000);
    setTimeout(() => {
      console.log(3, props.router.query);
    }, 10000);
```
What am I doing wrong? Btw, I'm getting the logs two times (Probably one time on static generation and one time when I'm going to the page), but the object is empty both times

**Reply by adebiyial** [reply]
I don't think using a timeout is ideal. Why don't you try it before you `return` and conditionally render.

**Reply by Paultje52** [reply]
Do you mean save the `props.router` object in the class (With `this.router`) and try it in the `render()` method?

**Reply by adebiyial** [reply]
Yes. Check for it in your render, not in the constructor.

```
render() {
  const { router } = this.props;
  console.log(router.query);

  // ...
}
```

**Reply by Paultje52** [reply]
Hi. I used the code you provided, but the query object is still empty. Is this a bug?

**Reply by adebiyial** [reply]
Weird. Can you send me a screenshot of your component?

**Reply by Paultje52** [reply]
Sure. Here is my whole code.

```js
import React, { Component } from "react";
import Router, { withRouter } from "next/router";
import WebSocket from "../../components/dashboard/WebSocket.js";
import styles from "../../styles/dashboard/Index.module.css";
import Navigation from "../../components/dashboard/Navigation.js";
import cookieCutter from "cookie-cutter";
import NotFound from "../404.js";

import Home from "../../components/dashboard/Home.js";
import ConfirmMail from "../../components/dashboard/ConfirmMail.js";

class Dashboard extends Component {

  constructor(props) {
    super(props);
    console.log(1, props.router.query);

    this.state = {
      loading: true,
      online: true,
    
      apiUrl: "http://localhost:8080",
      page: props.router.query.page || [],
      user: false
    }; 
  }

  setUser(user) {
    // if (!user.isVerified()) Router.replace("/dashboard/email-confirm");

    this.setState({
      user,
      loading: false
    });
  }

  async componentDidMount() {

    if (!navigator.onLine) {
      this.setState({
        online: false,
        loading: false
      });
      return;
    };

    let session = cookieCutter.get("session");
    if (!session) return Router.replace("/login");
  
    let res = await fetch(`${this.state.apiUrl}/signin/checkSession?session=${session}`);
    let json = await res.json();

    if (!json.active) return Router.replace("/login");
    
    this.setState({
      loading: true
    });
    this.websocket = new WebSocket(this.state.apiUrl, session, this);
    
  }

  getCorrectPage() {
    switch(this.state.page[0]) {
      case "home":
        return (
          <Home
            dashboard={this}
          />
        );
      
      case "email-confirm":
        return (
          <ConfirmMail
            dashboard={this} 
          />
        )
      
      default:
        return (
          <NotFound
            dashboard={true}
          />
        );
    }
  }

  render() {

    const { router } = this.props;
    console.log(router.query);

    if (this.state.loading) {
      return (
        <div>
          <h1>Loading...</h1>
        </div>
      );
    }

    // TODO: Offline stuff!
    if (!this.loading && !this.online) {
      // let 
    }

    return (
      <div style={{height: "100%", marginTop: "-22px"}}>

        <Navigation
          user={this.state.user}
          path={this.state.page}
        />
        
        <div className={styles.mainContent}>
          {this.getCorrectPage()}
        </div>

      </div>
    )
  }

}

export default withRouter(Dashboard);
```

**Reply by adebiyial** [reply]
I tried it. It works. Let's see what's in your console. Log `router` alone in this case.
```
const { router } = this.props;
console.log(router);
```

Send a screenshot of your browser console log.

<img width="462" alt="image" src="https://user-images.githubusercontent.com/18689106/122650490-2cc77480-d12b-11eb-8862-434250ec079b.png">
Works for me.

**Reply by Paultje52** [reply]
I got it working this time too. Thank you for your help!

**Reply by adebiyial** [reply]
Ah. Great. 🎊.

Hmm. Is there a reason why you're still on class components?

**Reply by Paultje52** [reply]
I'm new to react/next.js and working with functions seems a lot harder than working with classes for me when needing to rerender, keep and share variables, etc. I'm just more familiar with classes.

Anyways, I'm trying to get `next/image` to work, but I'm getting an error. Can you maybe help me with that too? This is my code.
```js
import Image from "next/image";
import bannerImg from "../../public/banner.png"

export default function Navigation({ user, path, gotoPage }) {

  return (
    <div className={styles.leftNav}>
      <Image src={bannerImg} />
      
    </div>
  );

}

```

This is my error.
```
./public/banner.png 1:0
Module parse failed: Unexpected character '�' (1:0)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
(Source code omitted for this binary file)
```

**Reply by adebiyial** [reply]
Oh. Great. Functions are a different thing.

Inline the image src like

```
<Image src="/banner.png" alt="..."/>
```

**Reply by Paultje52** [reply]
Hey! Thank tou for your response. I think the docs are off, that's why I thought I had to import the image. Anyways, is there a way to style the image? Adding `className` doesn't work.
```jsx
<Image src={"/banner.png"} className={navCss.banner} layout="fill" />
```

**Reply by adebiyial** [reply]
I don't think so because what you have there is a component you don't control. What you can do is create a wrapper element like a `div`, attach the class there, and target the `img` accordingly.

**Reply by Paultje52** [reply]
I tried it, but it doesn't work. Here is my jsx.
```jsx
      <div className={navCss.banner}>
        <Image src={"/banner.png"} layout="fill" />
      </div>
```
And here is my css.
```css
.banner img {
  top: 0;
  left: 0;
  margin-top: 25px;
  max-width: 100%;
}
```

This is what it's supposed to be (Done with a normal img tag).
![image](https://user-images.githubusercontent.com/31313717/122682326-ffe19300-d1f8-11eb-94ea-1b96980449d4.png)

But this is what I see.
![image](https://user-images.githubusercontent.com/31313717/122682347-1851ad80-d1f9-11eb-8f93-5698685c4eb3.png)

**Reply by adebiyial** [reply]
What you need to do is inspect element in your browser and see if the `img` is been targeted. And also check if the class names are been properly attached to the necessary elements.

**Reply by Paultje52** [reply]
I'm sure that it's correct, because the only thing I changed was the element name, from `Image` to `img` and then everything was correct

**Reply by adebiyial** [reply]
But you have to debug through the DOM. At least check to see what's been rendered.

**Reply by Paultje52** [reply]
![image](https://user-images.githubusercontent.com/31313717/122683572-5b634f00-d200-11eb-99bd-58506370f1a0.png)
![image](https://user-images.githubusercontent.com/31313717/122683576-661de400-d200-11eb-874b-ba1580538b39.png)

**Reply by adebiyial** [reply]
So what's happening is that `next/image' has some styles that is overriding yours. `inset` overrides your `top` and `left`. `margin` cancels `margin-top`.

My verdict is that you let `next/image` take care of those styles so it can work as intended, or if you need to override it's styles, then you can apply `!important` to your own styles.

Or use the Inline style on `Image`
